### PR TITLE
IO4 - Usable commands factory

### PR DESCRIPTION
**Pull request by startled-cat**

### CODE REVIEW

**Comment by Luvkitri** [comment]
Seems about right

File: src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java [hunk 1]
```diff
@@ -97,7 +98,35 @@ public CommandManagerWindow(DriverCommandManager commandManager) {
...
+			commandManager.commandFactory.clear();
+			cfChoice.removeAll();
+		});
+		btnFactorySetCommand.addActionListener(e -> {
+			commandManager.setCurrentCommand(cfChoice.getSelectedItem());
+		});
+
+		content.add(btnFactoryClearCommand, c);
+		content.add(btnFactorySetCommand, c);
+
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java]: Extract factory gui configuration to a separate method.

File: src/main/java/edu/kis/powp/jobs2d/command/manager/DriverCommandManager.java [hunk 2]
```diff
@@ -37,8 +41,27 @@ public DriverCommandManager(DriverManager driverManager) {
 	 * @param name        name of the command.
 	 */
 	public synchronized void setCurrentCommand(List<DriverCommand> commandList, String name) {
+		commandFactory.add(name, new DefaultCompoundCommand((commandList)));
+		currentCommandName = name;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/DriverCommandManager.java]: OK. Actually 'name' should be added to the DefaultCompoundCommand (and it should be returned by toString)

File: src/main/java/edu/kis/powp/jobs2d/command/manager/DriverCommandManager.java [hunk 3]
```diff
@@ -37,8 +41,27 @@ public DriverCommandManager(DriverManager driverManager) {
 	 * @param name        name of the command.
 	 */
 	public synchronized void setCurrentCommand(List<DriverCommand> commandList, String name) {
+		commandFactory.add(name, new DefaultCompoundCommand((commandList)));
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/DriverCommandManager.java]: No. You should not add the dependency to the commandFactory here. Use observer (see publisher)! This line should be implemented in a separate commandManager subscriber implementation.

File: src/main/java/edu/kis/powp/jobs2d/command/manager/DriverCommandManager.java [hunk 4]
```diff
@@ -37,8 +41,27 @@ public DriverCommandManager(DriverManager driverManager) {
...
 		setCurrentCommand(new DefaultCompoundCommand(commandList));
 	}
+	/**
+	 * Sets current command.
+	 *
+	 * @param name        name of the command.
+	 */
+	public synchronized void setCurrentCommand(String name) {
+		setCurrentCommand(commandFactory.get(name));
+	}
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/DriverCommandManager.java]: This also should be removed. You can call setCurrentCommand from the factory (without creating reverse dependency)

File: src/main/java/edu/kis/powp/jobs2d/features/CommandFactory.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import edu.kis.powp.jobs2d.command.ImmutableCompoundCommand;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+
+public class CommandFactory {
+
+    private HashMap<String, ICompoundCommand> shapes;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/features/CommandFactory.java]: The declaration should indicate only the interface, i.e. Map.
I think that the type of values should also be more general, i.e., DriverCommand

File: src/main/java/edu/kis/powp/jobs2d/features/CommandFactory.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        shapes = new HashMap<>();
+    }
+
+    public ICompoundCommand get(String name){
+        return shapes.get(name);
+    }
+
+    public void add(String name, ICompoundCommand commands) throws IllegalArgumentException{
+        try{
+            shapes.put(name, (DefaultCompoundCommand) commands.clone());
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/features/CommandFactory.java]: Is casting to DefaultCompoundCommand really important?

File: src/test/java/edu/kis/powp/jobs2d/features/CommandFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        cf.get("line").forEach(listFromfactory::add);
+
+        assertEquals(listOriginal.size(), listFromfactory.size());
+
+        for (int i = 0; i < listOriginal.size(); i++) {
+            if(listOriginal.get(i) instanceof OperateToCommand){
+                assert(listFromfactory.get(i) instanceof OperateToCommand);
+                OperateToCommand a1 = (OperateToCommand)listOriginal.get(i);
+                OperateToCommand a2 = (OperateToCommand)listOriginal.get(i);
+                assertEquals(a1.getX(), a2.getX());
```
Comment (kowallus) [inline src/test/java/edu/kis/powp/jobs2d/features/CommandFactoryTest.java:89]: How could possibly a1 differ from a2 if they are both listOriginal.get(i)?! This part of the test does not make sense.

File: src/test/java/edu/kis/powp/jobs2d/features/CommandFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    @Test
+    void getNamesOfStored() {
+        CommandFactory cf = new CommandFactory();
+
+        List<DriverCommand> dcList = new ArrayList<>();
+        dcList.add(new SetPositionCommand(100, 100));
+        dcList.add(new OperateToCommand(150, 150));
+        ImmutableCompoundCommand icc = new ImmutableCompoundCommand(dcList);
+
+        cf.add("line", icc);
```
Comment (kowallus) [inline src/test/java/edu/kis/powp/jobs2d/features/CommandFactoryTest.java]: You should better use constants.

**Review by kowallus — CHANGES_REQUESTED** [review]
Some minor issues and to many changes in DriverCommandManager makes merging impossible.

**Comment by kowallus** [comment]
please close, fix and reopen PR

**Pull request merged**